pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -122,26 +122,344 @@ extension type Tool.fromMap(Map<String, Object?> _value) {
     List<String>? required,
+
+    /// Must be one of bool, Schema, or Null
+    Object? additionalProperties,
+    bool? unevaluatedProperties,
+    StringSchema? propertyNames,
+    int? minProperties,
+    int? maxProperties,
   }) =>
-      InputSchema.fromMap({
-        'type': 'object',
+      ObjectSchema.fromMap({
+        'type': JsonType.object.typeName,
+        if (title != null) 'title': title,
+        if (description != null) 'description': description,
         if (properties != null) 'properties': properties,
+        if (patternProperties != null) 'patternProperties': patternProperties,
         if (required != null) 'required': required,
+        if (additionalProperties != null)
+          'additionalProperties': additionalProperties,
+        if (unevaluatedProperties != null)
+          'unevaluatedProperties': unevaluatedProperties,
+        if (propertyNames != null) 'propertyNames': propertyNames,
+        if (minProperties != null) 'minProperties': minProperties,
+        if (maxProperties != null) 'maxProperties': maxProperties,
       });
 
-  String get type => _value['type'] as String;
+  /// A map of the properties of the object to the nested [Schema]s for those
+  /// properties.
+  Map<String, Schema>? get properties =>
+      (_value['properties'] as Map?)?.cast<String, Schema>();
 
-  Map<String, Object?>? get properties =>
-      (_value['properties'] as Map?)?.cast<String, Object?>();
+  /// A map of the property patterns of the object to the nested [Schema]s for
+  /// those properties.
+  Map<String, Schema>? get patternProperties =>
+      (_value['patternProperties'] as Map?)?.cast<String, Schema>();
 
+  /// A list of the required properties by name.
   List<String>? get required => (_value['required'] as List?)?.cast<String>();
+
+  /// Rules for additional properties that don't match the
+  /// [properties] or [patternProperties] schemas.
+  ///
+  /// Can be either a [bool] or a [Schema], if it is a [Schema] then additional
+  /// properties should match that [Schema].
+  /*bool|Schema|Null*/ Object? get additionalProperties =>
+      _value['additionalProperties'];
+
+  /// Similar to [additionalProperties] but more flexible, see
+  /// https://json-schema.org/understanding-json-schema/reference/object#unevaluatedproperties
+  bool? get unevaluatedProperties => _value['unevaluatedProperties'] as bool?;
+
+  /// A list of valid patterns for all property names.
+  StringSchema? get propertyNames =>
+      (_value['propertyNames'] as Map?)?.cast<String, Object?>()
+          as StringSchema?;
+
+  /// The minimum number of properties in this object.
+  int? get minProperties => _value['minProperties'] as int?;
+
+  /// The maximum number of properties in this object.
+  int? get maxProperties => _value['maxProperties'] as int?;
+}
+
+/// A JSON Schema definition for a String.
+extension type const StringSchema.fromMap(Map<String, Object?> _value)
+    implements Schema {
+  factory StringSchema({
```

Comment:
Can extension types have forwarding constructors? I think `Schema.string(...)` is a friendly API than `StringSchema(...)` because it has nice autocomplete after `Schema.`

Reply:
The main thing is you want to get the proper subtype of out it.

I could use static methods to achieve that API though.

Reply:
I think if I did that approach I would also want to remove the regular constructors from the subtypes, so I don't have to duplicate them all?